code/modules/mafia/controller.dm

```diff
@@ -667,31 +697,81 @@
 		. += L[key]
 
 /**
-  * Returns all setups that the amount of players signed up could support (so fill each role)
-  * Arguments:
-  * * ready_count: the amount of players signed up (not sane, so some players may have disconnected or rejoined ss13).
+  * Returns a semirandom setup, with...
+  * Town, Two invest roles, one protect role, sometimes a misc role, and the rest assistants for town.
+  * Mafia, 2 normal mafia and one special.
+  * Neutral, two disruption roles, sometimes one is a killing.
+  *
+  * Assistants must be filled in their own list before merging the two because of how byond dicts work
+  * See _defines.dm in the mafia folder for a rundown on what these groups of roles include.
+  */
+/datum/mafia_controller/proc/generate_random_setup()
+	var/invests_left = 2
+	var/protects_left = 1
+	var/miscs_left = prob(35) ? 1 : 0
+	var/mafiareg_left = 2
+	var/mafiaspe_left = 1
+	var/killing_role = prob(50) ? 1 : 0
+	var/disruptors = killing_role ? 1 : 2 //still required to calculate overflow
+
+	var/overflow = 12 - (invests_left + protects_left + miscs_left + mafiareg_left + mafiaspe_left + killing_role + disruptors)
+	var/list/assistants = list(/datum/mafia_role = overflow) //amount of people who are not any of the special roles
+	var/list/random_setup = list()
+	for(var/i in 1 to (12 - overflow)) //should match the number of roles to add
+		if(invests_left)
+			add_setup_role(random_setup, TOWN_INVEST)
+			invests_left--
+		else if(protects_left)
+			add_setup_role(random_setup, TOWN_PROTECT)
+			protects_left--
+		else if(miscs_left)
+			add_setup_role(random_setup, TOWN_MISC)
+			miscs_left--
+		else if(mafiareg_left)
+			add_setup_role(random_setup, MAFIA_REGULAR)
+			mafiareg_left--
+		else if(mafiaspe_left)
+			add_setup_role(random_setup, MAFIA_SPECIAL)
+			mafiaspe_left--
+		else if(killing_role)
+			add_setup_role(random_setup, NEUTRAL_KILL)
+			killing_role--
+		else
+			add_setup_role(random_setup, NEUTRAL_DISRUPT)
```

Comment:
May want to commit to the else if here just for clarity

Reply:
we'll see what maintainers say because i'm sure they have some choice words for this proc in general. i could see it either way